consensus/istanbul/backend/randao.go

```diff
@@ -58,12 +61,32 @@ func (p *ChainBlsPubkeyProvider) getAllCached(chain consensus.ChainReader, num *
 	}
 
 	backend := backends.NewBlockchainContractBackend(chain, nil, nil)
-	kip113Addr, err := system.ReadRegistryActiveAddr(backend, system.Kip113Name, num)
-	if err != nil {
-		return nil, err
+	if common.Big0.Cmp(num) == 0 {
+		return nil, errors.New("num cannot be zero")
+	}
+	parentNum := new(big.Int).Sub(num, common.Big1)
+
+	var kip113Addr common.Address
+	// Because the system contract Registry is installed at Finalize() of RandaoForkBlock,
+	// it is not possible to read KIP113 address from the Registry at RandaoForkBlock.
+	// Hence the ChainConfig fallback.
+	if chain.Config().IsRandaoForkBlock(num) {
+		var ok bool
+		kip113Addr, ok = chain.Config().RandaoRegistry.Records[system.Kip113Name]
+		if !ok {
+			return nil, errors.New("KIP113 address not set in ChainConfig")
+		}
+	} else if chain.Config().IsRandaoForkEnabled(num) {
+		var err error
+		kip113Addr, err = system.ReadRegistryActiveAddr(backend, system.Kip113Name, parentNum)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, errors.New("Cannot read KIP113 address from registry before Randao fork")
```

Comment:
`ChainBlsPubkeyProvider` should be used only after Randao HF, Randao-disabled condition is not needed 

Prerequisites of Randao Hardfork 
- BLS contract (KIP113) should be deployed
- Validators information should be registered on the BLS contract
- Randao registry should be specified with the KIP113 contract address 

```suggestion
	} else {
		var err error
		kip113Addr, err = system.ReadRegistryActiveAddr(backend, system.Kip113Name, parentNum)
		if err != nil {
			return nil, err
		}
```

Reply:
You're right. I wasn't sure what would happen when this function is invoked before Randao.

Reply:
If the function is called before Randao HF, it will fail at `system.ReadRegistryActiveAddr()` because the registry is registered at Randao HF. It is not an important issue, though.